alibi_detect/base.py

```diff
@@ -93,40 +93,89 @@ def infer_threshold(self, X: np.ndarray) -> None:
+        -------
+        The detector's configuration dictionary.
+        """
+        if self.config is not None:
+            # Get config (stored in top-level self)
+            cfg = self.config
+            # Get low-level nested detector (if needed)
+            detector = self._detector if hasattr(self, '_detector') else self  # type: ignore[attr-defined]
+            detector = detector._detector if hasattr(detector, '_detector') else detector  # type: ignore[attr-defined]
+            # Add large artefacts back to config
+            for key in LARGE_ARTEFACTS:
+                if key in cfg:  # self.config is validated, therefore if a key is not in cfg, it isn't valid to insert
+                    cfg[key] = getattr(detector, key)
+            # Set x_ref_preprocessed flag
+            preprocess_at_init = getattr(detector, 'preprocess_at_init', True)  # If no preprocess_at_init, always true!
+            cfg['x_ref_preprocessed'] = preprocess_at_init and detector.preprocess_fn is not None
+            return cfg
+        else:
+            raise NotImplementedError('Getting a config (or saving via a config file) is not yet implemented for this'
+                                      'detector')
+
+    @classmethod
+    def from_config(cls, config: dict):
+        """
+        Instantiate a drift detector from a fully resolved (and validated) config dictionary.
+
+        Parameters
+        ----------
+        config
+            A config dictionary matching the schema's in :class:`~alibi_detect.saving.schemas`.
+        """
+        # Check for exisiting version_warning. meta is pop'd as don't want to pass as arg/kwarg
+        version_warning = config.pop('meta', {}).pop('version_warning', False)
+        # Init detector
+        detector = cls(**config)
+        # Add version_warning
+        detector.meta['version_warning'] = version_warning  # type: ignore[attr-defined]
+        detector.config['meta']['version_warning'] = version_warning
+        return detector
+
+    def _set_config(self, inputs):  # TODO - move to BaseDetector once config save/load implemented for non-drift
+        # Set config metadata
         name = self.__class__.__name__
         # strip off any backend suffix
         backends = ['TF', 'Torch', 'Sklearn']
         for backend in backends:
             if name.endswith(backend):
                 name = name[:-len(backend)]
         # Init config dict
-        cfg: Dict[str, Any] = {'name': name}
-
-        # Add x_ref
-        cfg.update({'x_ref': self.x_ref})
+        self.config: Dict[str, Any] = {
+            'name': name,
+            'meta': {
+                'version': self.meta['version'],
+                'config_spec': __config_spec__,
+            }
+        }
 
-        # Add preprocess_fn field
-        if self.preprocess_fn is not None:
-            cfg.update({'preprocess_fn': self.preprocess_fn})
+        # args and kwargs
+        pop_inputs = ['self', '__class__', '__len__']
```

Comment:
Where do the `__class__` and `__len__` come from? Is this an exhaustive list of things to get rid of?

Reply:
Yep. Because `inputs` is a dict, it has `self`, `__class__` and `__len__` attributes. We want to strip these so that we're left with only the args and kwargs. Also do `pop_inputs += self.config.keys()` so as to avoid overwriting existing args/kwargs.

Reply:
I'm confused, a `locals()` `dict` has many more attributes and methods than these, I'm not sure what's so special about e.g. `__len__`?

Reply:
ok interesting...